energy-integration/governance-v2/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+            description,
+            actions: gov_actions,
+        };
+        let proposal_id = self.proposals().push(&proposal);
+
+        if !payments_for_action.is_empty() {
+            self.pending_payments_for_proposal(proposal_id)
+                .set(&payments_for_action);
+        }
+
+        self.total_votes(proposal_id).set(&user_energy);
+        let _ = self.user_voted_proposals(&proposer).insert(proposal_id);
+
+        let current_block = self.blockchain().get_block_nonce();
+        self.proposal_start_block(proposal_id).set(&current_block);
+
+        self.proposal_created_event(proposal_id, &proposer, current_block, &proposal);
+
+        proposal_id
+    }
+
+    /// Vote on a proposal. The voting power depends on the user's energy.
+    #[endpoint]
+    fn vote(&self, proposal_id: ProposalId) {
+        self.require_caller_not_self();
+        self.require_valid_proposal_id(proposal_id);
+        require!(
+            self.get_proposal_status(proposal_id) == GovernanceProposalStatus::Active,
+            "Proposal is not active"
+        );
+
+        let voter = self.blockchain().get_caller();
+        let new_user = self.user_voted_proposals(&voter).insert(proposal_id);
+        require!(new_user, ALREADY_VOTED_ERR_MSG);
+
+        let user_energy = self.get_energy_non_zero(voter.clone());
+        self.total_votes(proposal_id)
+            .update(|total_votes| *total_votes += &user_energy);
+
+        self.vote_cast_event(&voter, proposal_id, &user_energy);
+    }
+
+    /// Downvote a proposal. The voting power depends on the user's energy.
+    #[endpoint]
+    fn downvote(&self, proposal_id: ProposalId) {
+        self.require_caller_not_self();
+        self.require_valid_proposal_id(proposal_id);
+        require!(
+            self.get_proposal_status(proposal_id) == GovernanceProposalStatus::Active,
+            "Proposal is not active"
+        );
+
+        let downvoter = self.blockchain().get_caller();
+        let new_user = self.user_voted_proposals(&downvoter).insert(proposal_id);
+        require!(new_user, ALREADY_VOTED_ERR_MSG);
+
+        let user_energy = self.get_energy_non_zero(downvoter.clone());
+        self.total_downvotes(proposal_id)
+            .update(|total_downvotes| *total_downvotes += &user_energy);
+
+        self.downvote_cast_event(&downvoter, proposal_id, &user_energy);
+    }
+
+    /// Queue a proposal for execution.
+    /// This can be done only if the proposal has reached the quorum.
+    /// A proposal is considered successful and ready for queing if
+    /// total_votes - total_downvotes >= quorum
+    /// and all the required payments were deposited
+    #[endpoint]
+    fn queue(&self, proposal_id: ProposalId) {
+        self.require_caller_not_self();
+        require!(
+            self.get_proposal_status(proposal_id) == GovernanceProposalStatus::Succeeded,
+            "Can only queue succeeded proposals"
+        );
+        require!(
```

Comment:
I think this condition must be updated.
There should not be a condition for self.pending_payments_for_proposal. When you make a new proposal you set this value and you clear it only in the clearProposal endpoint. Furthermore, I think we should not have this condition because what happens if a proposal requires no payments, as stated in the check from the depositTokensForProposal endpoint?
Also, if there is a need for a second check, besides self.payments_depositor, I think we should have an AND check (&&) instead of OR (||).